File: Exams/Resoluções de Exames by Carlos Martins/time-process-thread-synch/synch/Graph.java
/***
 *
 * ISEL, LEIC, Concurrent Programming
 *
 * Control synchronization in Java
 *
 * Carlos Martins, October 2016
 *
 ***/

/*
The action graph is as follows:

- Starts with A;
- After A finishes execution, can be executed B, C, D and E;
- After D and E finish execution, can be executed F;
- After B, C and F finish execution, can be executed G.

In this design:
- Thread T1 executes A, B and G;
- Thread T2 executes C;
- Thread T3 executes D and F.
- Thread T4 executes E.
*/

import java.util.Random;
import java.util.concurrent.*;
import java.util.concurrent.locks.*;

public class Graph {

	//
	// Synchronizers
	//
	// The CountDownLatch when initialized with 1 has a synchronization semantics
	// equivalent to a manual-reset event (without Reset operation).
	//

	private static final CountDownLatch t1_2_t2t3t4 = new CountDownLatch(1);
	private static final CountDownLatch t4_2_t3 = new CountDownLatch(1);
	private static final Semaphore t2t3_2_t1 = new Semaphore(0, false);

	//
	// The worker threads
	//

	private static Thread thr1, thr2, thr3, thr4;

	//
	// Constants used to generate random action durations.
	//

	private static final int MIN_DURATION = 50;
	private static final int MAX_DURATION = 2000;

	//
	// Execute the specified action.
	//
	
	private static void doAction(String actionName, Random rand) {
		boolean interrupted = false;
		
		System.out.print("\n-->" + actionName);
		for (;;)
			try {
				Thread.sleep(rand.nextInt(MAX_DURATION) + MIN_DURATION);
				break;
			} catch (InterruptedException ie) {
				interrupted = true;
			}
		System.out.print("\n<--" + actionName);
		
		// If the current thread was interrupted, re-assert the interruption.
		if (interrupted)
			Thread.currentThread().interrupt();
	}
	
	//
	// Await uninterruptibly on a CountDownLatch.
	//
	
	private static void awaitUninterruptibly(CountDownLatch cdl) {
		boolean interrupted = false;
		
		do {
			try {
				cdl.await();
				break;
			} catch (InterruptedException ie) {
				interrupted = true;
			}
		} while (true);
			
		// If the current thread was interrupted, re-assert the interruption.
		if (interrupted)
			Thread.currentThread().interrupt();
	}

	//
	// Join with a thread uninterruptibly.
	//
	
	private static void joinUninterruptibly(Thread t) {
		boolean interrupted = false;
		
		do {
			try {
				t.join();
				break;
			} catch (InterruptedException ie) {
				interrupted = true;
			}
		} while (true);
			
		// If the current thread was interrupted, re-assert the interruption.
		if (interrupted)
			Thread.currentThread().interrupt();
	}
	
	//
	// Thread T1
	//

	static class T1 implements Runnable {
		public void run() {
			Random rand = new Random(Thread.currentThread().getId() << 8);
		
			// Do action A
			doAction("A", rand);
	
			// Allow thread T2, T3 and T4 start executing.
			t1_2_t2t3t4.countDown();
	
			// Do action B
			doAction("B", rand);

			// Wait until T2 and T3 execute completely.

			t2t3_2_t1.acquireUninterruptibly(2);
	
			// Do action G
			doAction("G", rand);
		}
	}

	//
	// Thread T2
	//

	private static class T2 extends Thread {
		public void run() {
			Random rand = new Random(Thread.currentThread().getId() << 9);
	
			// Wait until T1 completes execution of action A
			awaitUninterruptibly(t1_2_t2t3t4);
	
			// Do action C
			doAction("C", rand);
			
			// Signals T1
			t2t3_2_t1.release();
		}
	}

	//
	// Thread T3
	//

	private static class T3 extends Thread {
		public void run() {
			Random rand = new Random(Thread.currentThread().getId() << 10);
	
			// Wait until t1 completes the action A
			awaitUninterruptibly(t1_2_t2t3t4);
	
			// Do action D
			doAction("D", rand);

			// Wait until the T4 completes execution of action E
			awaitUninterruptibly(t4_2_t3);
	
			// Do action F
			doAction("F", rand);
			
			// Signals T1 in order to allow execution of action G
			t2t3_2_t1.release();
		}
	}

	//
	// Thread T4
	//

	private static class T4 extends Thread {
		public void run() {
			Random rand = new Random(Thread.currentThread().getId() << 11);
		
			// Wait until T1 completes execution of action A
			awaitUninterruptibly(t1_2_t2t3t4);
	
			// Do action E
			doAction("E", rand);

			// Signal T3 in order to allw execution of action F	
			t4_2_t3.countDown();
		}
	}

	public static void main(String[] args) {

		// Create the executer thread objects.
		// T1 is executed by the primary thread.
		thr1 = Thread.currentThread();
		thr2 = new T2();
		thr3 = new T3();
		thr4 = new T4();
		
		System.out.print("++executing actions...");

		// Start the executer threads
		thr2.start();
		thr3.start();
		thr4.start();
		
		// T1 is executed by the primary thread.
		new T1().run();
		// T1 executes the last Action, so print newline
		System.out.println();
	
		//
		// Since all thread are foreground threads, the process
		// only exits when all threads have been exited.
		// 	
	}
}

File: Exams/Resoluções de Exames by Carlos Martins/time-process-thread-synch/synch/Graph.cs
/***
 *
 * ISEL, LEIC, Concurrent Programming
 *
 * Control synchronization in .NET
 *
 * Carlos Martins, October 2016
 *
 ***/



/*
The action graph is as follows:

- Starts with A;
- After A finishes execution, can be executed B, C, D, and E;
- After D and E finish execution, can be executed F;
- After B, C and F finish execution, can be executed G.

In this design:
- Thread T1 executes A, B and G;
- Thread T2 executes C;
- Thread T3 executes D and F;
- Thread T4 executes E.
*/
		
//
// Uncomment the next define in order to user THREAD_JOIN
//

//#define USE_THREAD_JOIN


using System;
using System.Threading;

static class Graph {

	//
	// Synchronizers
	//

	private static ManualResetEventSlim t1_2_t2t3t4 = new ManualResetEventSlim(false);

#if (!USE_THREAD_JOIN)

	// In this usage CountdownEvent is similar to ManualResetEventSlim!
	private static CountdownEvent t4_2_t3 = new CountdownEvent(1);
	
	private static SemaphoreSlim t2t3_2_t1 = new SemaphoreSlim(0, 2);

#endif

	//
	// Thread references
	//

	private static Thread[] ths = new Thread[4];

	//
	// Sleep constants
	//

	private static int MIN_TIMEOUT = 50;
	private static int MAX_TIMEOUT = 2000;

	//
	// Simulate action execution.
	//
	
	private static void DoAction(string actionName, int duration)
	{
		Console.Write("\n-->" + actionName);
		Thread.Sleep(duration);
		Console.Write("\n<--" + actionName);
	}
	
	//
	// Thread T1
	//

	private static void T1()
	{
		Random rand = new Random(Thread.CurrentThread.ManagedThreadId);
		
		// Do action A
		DoAction("A", rand.Next(MIN_TIMEOUT, MAX_TIMEOUT));
	
		// Allow T2, T3 and T4 to start.
	
		t1_2_t2t3t4.Set();
	
		// Do action B
		DoAction("B", rand.Next(MIN_TIMEOUT, MAX_TIMEOUT));

#if (USE_THREAD_JOIN)

		// Wait until threads T2 and T3 have been exited.
		ths[1].Join();
		ths[2].Join();
#else	

		// Acquire two permissions from the semaphore.

		t2t3_2_t1.Wait();
		t2t3_2_t1.Wait();
#endif
	
		// Do action G
		DoAction("G", rand.Next(MIN_TIMEOUT, MAX_TIMEOUT));
	}

	//
	// Thread T2
	//

	private static void T2()
	{
		Random rand = new Random(Thread.CurrentThread.ManagedThreadId);
	
		// Wait until T1 completes A
		t1_2_t2t3t4.Wait();	
	
		// Do action C
		DoAction("C", rand.Next(MIN_TIMEOUT, MAX_TIMEOUT));

#if (!USE_THREAD_JOIN)
	
		t2t3_2_t1.Release();
#endif
	}

	//
	// Thread T3
	//

	private static void T3()
	{
		Random rand = new Random(Thread.CurrentThread.ManagedThreadId);
	
		// Wait until t1 completes A
		t1_2_t2t3t4.Wait();	
	
		// Do action D
		DoAction("D", rand.Next(MIN_TIMEOUT, MIN_TIMEOUT));

		// Wait until T4 completes action E
	
#if (USE_THREAD_JOIN)

		ths[3].Join();	
#else

		// Wait until the t4_2_t3 event is signalled
		t4_2_t3.Wait();	
#endif
	
		// Do action F
		DoAction("F", rand.Next(MIN_TIMEOUT, MAX_TIMEOUT));

#if (!USE_THREAD_JOIN)
	
		t2t3_2_t1.Release();
#endif
	}

	//
	// Thread T4
	//

	private static void T4()
	{
		Random rand = new Random(Thread.CurrentThread.ManagedThreadId);
		
		// Wait until T1 completes action A
		t1_2_t2t3t4.Wait();	
	
		// Do action E
		DoAction("E", rand.Next(MIN_TIMEOUT, MAX_TIMEOUT));

#if (!USE_THREAD_JOIN)
	
		t4_2_t3.Signal();
#endif
	}


	private static void Main()
	{

		// Create the worker threads.
		ths[0] = Thread.CurrentThread;
		ths[1] = new Thread(T2);
		ths[2] = new Thread(T3);
		ths[3] = new Thread(T4);

		// Start the threads
		ths[1].Start();
		ths[2].Start();
		ths[3].Start();

		// Wait for a while
		Thread.Sleep(500);
		
		Console.Write("++executing actions...");

		
		// T1 is executed by the primary thread.
		T1();
		//Console.WriteLine();
		//
		// Since all thread are foreground threads, the process only exits
		// when all threads have been exited.
		// 	
	}
}
